xmtp_mls/src/groups/mls_sync.rs

```diff
@@ -1682,63 +1712,91 @@ fn extract_message_sender(
+    let membership_diff = old_group_membership.diff(new_group_membership);
 
-    let old_group_membership = extract_group_membership(&extensions)?;
-    let new_group_membership = intent_data.apply_to_group_membership(&old_group_membership);
-
-    // Diff the two membership hashmaps getting a list of inboxes that have been added, removed, or updated
-    let membership_diff = old_group_membership.diff(&new_group_membership);
-
-    // Construct a diff of the installations that have been added or removed.
-    // This function goes to the network and fills in any missing Identity Updates
     let installation_diff = client
         .get_installation_diff(
             provider.conn_ref(),
-            &old_group_membership,
-            &new_group_membership,
+            old_group_membership,
+            new_group_membership,
             &membership_diff,
         )
         .await?;
 
-    let mut new_installations: Vec<Installation> = vec![];
-    let mut new_key_packages: Vec<KeyPackage> = vec![];
+    let mut new_installations = Vec::new();
+    let mut new_key_packages = Vec::new();
+    let mut failed_installations = Vec::new();
 
     if !installation_diff.added_installations.is_empty() {
-        let my_installation_id = &client.context().installation_public_key().to_vec();
-        // Go to the network and load the key packages for any new installation
+        let my_installation_id = client.context().installation_public_key().to_vec();
+
         let key_packages = client
             .get_key_packages_for_installation_ids(
                 installation_diff
                     .added_installations
-                    .into_iter()
-                    .filter(|installation| my_installation_id.ne(installation))
+                    .iter()
+                    .filter(|installation| my_installation_id.ne(*installation))
+                    .cloned()
                     .collect(),
             )
             .await?;
 
-        for key_package in key_packages {
-            // Add a proposal to add the member to the local proposal queue
-            new_installations.push(Installation::from_verified_key_package(&key_package));
-            new_key_packages.push(key_package.inner);
+        tracing::info!("trying to validate keypackages");
+
+        for (installation_id, result) in key_packages {
+            match result {
+                Ok(verified_key_package) => {
+                    new_installations.push(Installation::from_verified_key_package(
+                        &verified_key_package,
+                    ));
+                    new_key_packages.push(verified_key_package.inner.clone());
+                }
+                Err(_) => failed_installations.push(installation_id.clone()),
+            }
         }
     }
 
+    Ok(MembershipDiffWithKeyPackages::new(
+        new_installations,
+        new_key_packages,
+        installation_diff.removed_installations,
+        failed_installations,
```

Comment:
Does this get persisted somewhere on the group? Ideally we would keep track of the failed installations, so that they can be retried at some point later. Otherwise once this fails it will be very difficult for us to add retrying later, because we won't know who is missing.

Reply:
I would say it’s not our responsibility to re-add failed installations. There are two cases to consider:

1. All of a user’s installations fail – In this case, we inform the user that we can’t add them. They need to resolve the issue with their account before we can proceed.
2. At least one of the user’s installations fails, but not all – In this case, the successful installations should handle adding any missing installations.

With this approach, we place the responsibility on the user rather than the admin to retry. This is because there’s a chance that the problematic installation may never be fixed.

Reply:
I'm thinking about retrying at a much later date. For example, if you have an expired key package that got dropped from the network and then come back online and upload a new key package.

If we don't store the failed installations it requires a very expensive diffing operation to figure out who is missing from a large group.

Reply:
I'm still thinking of if we distribute the load on the clients of each user it would be better, then we can show their healthy clients some alerts that you have a bad installation go and fix it, or here you can revoke that installation!
Just showing the admin that we can't add that user bc they don't have a healthy client, then the admin can add them later rather than doing it in the background.

Reply:
I don't think we have a mechanism for adding the unhealthy client later if we don't store it.

The inbox's `sequence_id` will be set to the latest. Uploading a new key package won't change your `sequence_id`. So even if you try to add the user again it won't see any installations in the diff

Reply:
Yes, we need to implement it. this is my idea:

- the client fetch the list of all installations from either the server or the sync_group
- on every welcome message the client check if all of installations are in the group, if not, the tries to add
- on every membership change, we can do like what we do on welcome messages, just to make sure we have our installations always are updated
- based on the installations list, the client will fetch all the keypackages for other installations, if something wrong, it will inform the user-- we can provide a list of all the user installations and their status

Reply:
just to be clear, I 100% agree of saving the failed installations, just wanted to put the load on other clients.
Thanks for the comments Nick, as always insightful. will cover this part in the PR too 🙌

Reply:
FWIW, I don't think you need to handle the retrying now. We just need to know how it will work to make sure the right data is persisted to the group so we can retry it later.